### ISSUE TITLE
Header Key Not LowerCase

### ISSUE THREAD

**Issue by ekoeryanto** [fastify/fastify#1396]
Before you submit an issue we recommend you drop into the [Gitter community](https://gitter.im/fastify) or [Fastify Help](https://github.com/fastify/help) and ask any questions you have or mention any problems you've had getting started with Fastify.

**Please read this entire template before posting any issue. If you ignore these instructions
and post an issue here that does not follow the instructions, your issue might be closed,
locked, and assigned the `missing discussion` label.**

## 🚀 Feature Proposal

Do not case set header function to lower

## Motivation

Make it more standard

## Example
```ts
reply.header('Retry-After', 123)
```
Please provide an example for how this feature would be used.

**Comment by mcollina** [comment]
Does this have any net effect on real-world applications?

**Comment by cemremengu** [comment]
AFAIK headers are case insensitive according to [RFC](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2)

**Comment by jsumners** [comment]
Yes, header names are case insensitive. Thus it is best to have a consistent case within the framework and that is lowercase.

**Comment by cemremengu** [comment]
Given the discussion, I think this has been concluded. Closing.

**Comment by zsimo** [comment]
Hello, a real-world scenario in which Header keys case mismatch can be a problem is:
Introduce Fastify in an already running API service, having customers who make case sensitive checks on header keys.

By the way, a quick solution can be to use Node Http core methods to set headers:
``` js
function onSend (request, reply, payload, next) {
    reply.res.setHeader("Ciao", "hello");
    next();
}
```
In this way, the case will be maintened.

Thanks!

**Comment by tlhunter** [comment]
I have worked with systems in the past that were unable to understand response headers that didn't follow the exact case they expect. For example, I had wrote a server that provided a response header of `content-type` , but the client couldn't understood it as it was hard-coded to look for `Content-Type`.

In this case, the client didn't follow the second rule of the [Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle), the full thing reading "Be conservative in what you send, be liberal in what you accept". In this case the client certainly should have been able to understand the response. I was very annoyed at the team implementing the client.

Try as I might, I couldn't get the other team to update their client. However, in this case, I realized I had also violated the principle myself, namely the first rule. While most of the programmer world understands HTTP headers, only a subset of those programmers understand that the headers should be case insensitive. The remainder assume headers will follow the expected `Documented-Like-This-Everywhere` capitalization.

I would love to see Fastify modified to be more robust with the HTTP response headers it exposes. As it stands, an HTTP response from Fastify seem to use an inconsistent set of cases, containing headers like `Date` and `content-length`. I haven't had the time to dig into this yet, but I'm assuming it's unintentional, with some libraries providing `Date` and Fastify lower-casing all headers (even those intentionally set by the application using capital letters).

**Comment by mcollina** [comment]
TL;DR

I do not expect this to be changed anytime soon because it is following the HTTP spec and there are no known compatibility problem.

---

The way we handler headers in Fastify and Node.js in general is to have them in a key/value map. However, the same header could appear multiple times, and they can all have different casing due to spec requirements. We must detect duplicates.

Lowercasing all headers is needed to deduplicate them, it is _fast_ and easy to understand. Node.js `res.setHeader` work around this problem by maintaining a more complex data structure that keep the lowercased version as well as the target one: https://github.com/nodejs/node/blob/1e665095c874a9bd21e7c88ce62ade991989bcf5/lib/_http_outgoing.js#L572.

Lowercasing is spec compliant - if there is a client not following the HTTP/1.1 spec... the bug is there, not in Fastify.

The reason why you see multiple casing is that the headers Node.js produces are using capital letters, e.g `Date`.

**Comment by jsumners** [comment]
Also, I believe we use the headers as provided by Node core. And our outgoing headers are similarly handled.
